### ISSUE TITLE
Component::onEnter() not called if component added after parent node is running

### ISSUE THREAD

**Issue by rh101** [axys1/axys#866]
- axys version: All, with latest checked being https://github.com/axys1/axys/tree/f8c3d95176422960a9aa210bd61425432851c948
- devices test on: N/A
- developing environments
   - NDK version: r19c
   - Xcode version: 12.4
   - VS/msvctools version: Visual Studio 2019(16.11)/14.29.30133
   - cmake version: 

This issue exists in Cocos2d-x as well.

Steps to Reproduce:

1. Add a Component to a node at any point after `Node::onEnter()` is run (so outside of `Node::init()` etc), such as in `Node::update()`.
2. `Component::onEnter()` is never called.

I can't seem to find any documentation related to this, so I'm not sure if this is a bug or intended behaviour, but it's strange that it would be implemented this way.

The `Node::addComponent()` is implemented like this:
```
bool Node::addComponent(Component* component)
{
    // lazy alloc
    if (!_componentContainer)
        _componentContainer = new ComponentContainer(this);

    // should enable schedule update, then all components can receive this call back
    scheduleUpdate();

    return _componentContainer->add(component);
}
```

The only place `Component::onEnter()` is called is here, where the component must already exist in the list prior to `Node::onEnter()` being called:
```
void Node::onEnter()
{
...
    if (_componentContainer && !_componentContainer->isEmpty())
    {
        _componentContainer->onEnter();
    }

    _isTransitionFinished = false;

    for (const auto& child : _children)
        child->onEnter();

    this->resume();

    _running = true;
}
```

This is completely different to when using `Node::addChild()`, as the `addChild()` method will actually call `Node::onEnter()` if the parent node is already running (it checks the `_running` flag):

```
void Node::addChild(Node* child, int localZOrder, int tag)
{
...
    addChildHelper(child, localZOrder, tag, "", true);
}

void Node::addChildHelper(Node* child, int localZOrder, int tag, std::string_view name, bool setTag)
{
...
    if (_running)
    {
        child->onEnter();
...
    }
...
}
```

Can anyone think of a reason to not call `Component::onEnter()` as well if the node is already running?

**Comment by DelinWorks** [comment]
From what i think, components are considered entered when they enter a node, not a node containing them entering a node, because that's what we care about in the end, we're just components of that node so we don't care where that node ends up as long as we're inside it and our onEnter has been called, but nodes are at the same level so when they exit or enter in each other, their functions get triggered. hope that helps 🙏 

Or i maybe wrong but that's how i see it

**Comment by rh101** [comment]
@DelinWorks  I don't quite understand what you mean, and how it relates to the actual issue.

To break down my initial post comparing how Components and child nodes differ in behaviour:

* Component::onEnter() is called from the owner/parent Node::onEnter() as long as the component already exists in the parent node.
* Component::onEnter() is not called if the component is added *after* the parent Node::onEnter() has already been called.

Now, the behaviour is different to how child nodes work:

* Child Node::onEnter() is called from the parent Node::onEnter() if it already exists in the child list.
* Child Node::onEnter() is also called from `addChild()` if the parent node is already running, meaning *after* Node::onEnter() has already been called.

**Comment by DelinWorks** [comment]
I do see what you mean now, I've been using the component class for a while and it's pretty badly written, I'll try to make a reasonable fix for your issue.

**Comment by rh101** [comment]
> I'll try to make a reasonable fix for your issue.

That's not necessary, as the fix is quite obvious, and I can take care of it, but the main purpose of this is to answer the question of whether the current behaviour is intended or not.

**Comment by DelinWorks** [comment]
Yeah it's like what you said, it doesn't make sense as it is because the component::onEnter may never get called. so it should be checked outside the node's is running flag

**Comment by rh101** [comment]
Assuming this is a bug, and not intended behaviour, then I'll post up the PR to fix it.  If anyone here thinks that the current implementation is the intended behaviour, then please try to explain why.

**Comment by rh101** [comment]
Fixed in #867
